### PR TITLE
matcher: Implement CEL matcher

### PR DESCRIPTION
**Pull request by tyxia**
Implement CEL ([Common Expression Language](https://github.com/google/cel-spec)) based matcher. 

- It consist of `cel_input` extension which provides matching data input and `cel_matcher` which provides matching input matcher. 

- CEL matcher is plugged in via `custom_match`. `buildMatcherTree()` function in cel_matcher_test.cc has detailed steps of building such matcher tree. The simplified configuration example : 

```
matcher_list:
    matchers:
      predicate:
        single_predicate:
          input:
            typed_config:
              "@type": type.googleapis.com/xds.type.matcher.v3.HttpAttributesCelMatchInput
          custom_match:
             typed_config: 
              "@type" : type.googleapis.com/xds.type.matcher.v3.CelMatcher
 ```

- Release note added.
- Tested CEL expression with supported request/response attributes, both positive(matched) and negative (notMatched) cases.

Signed-off-by: Tianyu Xia [tyxia@google.com](mailto:tyxia@google.com)

### CODE REVIEW

File: source/extensions/matching/http/cel_input/cel_input.h
Comment (yanavlasov) [inline source/extensions/matching/http/cel_input/cel_input.h:42]: What about request trailers?

**Review by yanavlasov — COMMENTED** [review]
/wait-any

File: source/extensions/matching/http/cel_input/cel_input.h
Comment (tyxia) [reply source/extensions/matching/http/cel_input/cel_input.h:42]: My understanding is that is not supported in Envoy CEL per [code](https://github.com/envoyproxy/envoy/blob/547ef8858f71ce7e997b53eadf5fd75b32395ea6/source/extensions/filters/common/expr/evaluator.h#L48-L50) . @kyessenov please correct me if i am wrong.

We can add the support for that later as we are most interested in request header for the moment.

Comment (kyessenov) [reply source/extensions/matching/http/cel_input/cel_input.h:42]: Yes, there was no need for request trailers till now.

File: source/extensions/matching/input_matchers/cel_matcher/matcher.h
Comment (kyessenov) [inline source/extensions/matching/input_matchers/cel_matcher/matcher.h:34]: This might be more naturally defined in the input header?

File: source/extensions/matching/http/cel_input/cel_input.h
Comment (kyessenov) [inline source/extensions/matching/http/cel_input/cel_input.h:38]: I think it probably makes sense to return CEL activation with only StreamInfo. It's still useful. Does this path ever get hit at all in the prod code?

File: envoy/matcher/matcher.h
Comment (kyessenov) [inline envoy/matcher/matcher.h:36]: Can we avoid shared_ptr here? Is there a reason why matching data is copied?

File: source/extensions/matching/input_matchers/cel_matcher/matcher.cc
Comment (kyessenov) [inline source/extensions/matching/input_matchers/cel_matcher/matcher.cc:10]: Builders are expensive - they are essentially "compiler instances". They need to be defined and reused in the factories - see other examples how it's done.

File: source/extensions/matching/input_matchers/cel_matcher/matcher.cc
Comment (kyessenov) [inline source/extensions/matching/input_matchers/cel_matcher/matcher.cc:18]: Compiled expression reference data from the underlying protobuf - the protobuf must be stored along with the compiled expression.

File: source/extensions/matching/input_matchers/cel_matcher/matcher.cc
Comment (kyessenov) [inline source/extensions/matching/input_matchers/cel_matcher/matcher.cc:35]: It should RELEASE_ASSERT or add nullptr check later on if you want debug ASSERT.

**Review by kyessenov — CHANGES_REQUESTED** [review]
Initial comments.

File: source/extensions/matching/input_matchers/cel_matcher/matcher.cc
Comment (tyxia) [reply source/extensions/matching/input_matchers/cel_matcher/matcher.cc:10]: yea, I created it in factory in previous revision but changed it to sit along with expression in most recent revision to make them in one place.

I agreed it better to set in factory. Will update.

File: envoy/matcher/matcher.h
Comment (kyessenov) [reply envoy/matcher/matcher.h:36]: It would be faster and easier code with unique_ptr. I'm not sure if we can avoid shared - would need to track where the ownership is split.

File: source/extensions/matching/input_matchers/cel_matcher/matcher.cc
Comment (tyxia) [reply source/extensions/matching/input_matchers/cel_matcher/matcher.cc:18]: hmm, I did see the comment in CEL lib that expr must outlive the resulting CelExpression before. But i was thinking it should cause issue(like if builder not outlive CelExpression) if i don't do that. And it runs fine so far.

I agree we are better off to store it but I am curious where does expression reference the data?

File: source/extensions/matching/input_matchers/cel_matcher/matcher.cc
Comment (tyxia) [reply source/extensions/matching/input_matchers/cel_matcher/matcher.cc:35]: I think we want to avoid the assert in data plane except config path. Here is not the config path.

Actually I am on the fence of having this assert. As i described in comment, it should not trigger. But I think it is good to have for debugging in the future. Initially I add envoy_log, but code coverage complains as I cannot hit it.

File: source/extensions/matching/http/cel_input/cel_input.h
Comment (tyxia) [reply source/extensions/matching/http/cel_input/cel_input.h:38]: This path will get hit as my test `CelMatcherNoRequestAttributes` shows.

Will it get hit in production? probably not. But I think it doesn't hurt to have it.

File: envoy/matcher/matcher.h
Comment (tyxia) [reply envoy/matcher/matcher.h:36]: yea it would be slightly faster, but i think current code is more flexible. As the interface, if user don't want to copy, they can std::move. I actually used `unique_ptr` throughout my CEL development until I ran the test using test_utility.h.
If user want to the pattern like what test_utility does, interface just allow it.

**Comment by tyxia** [comment]
@kyessenov @yanavlasov  Thanks for review! PTAL. Thanks.

**Review by yanavlasov — APPROVED** [review]

**Comment by yanavlasov** [comment]
Waiting for @kyessenov approval

File: test/extensions/matching/input_matchers/cel_matcher/cel_matcher_test.h
Comment (kyessenov) [inline test/extensions/matching/input_matchers/cel_matcher/cel_matcher_test.h:10]: nit: it's more common to have constexpr absl::string_view in envoy. `inline` is also not necessary, compiler should infer it.

Comment (kyessenov) [reply test/extensions/matching/input_matchers/cel_matcher/cel_matcher_test.h:10]: also, typo in the name: Heade -> Header

File: test/extensions/matching/input_matchers/cel_matcher/cel_matcher_test.h
Comment (kyessenov) [inline test/extensions/matching/input_matchers/cel_matcher/cel_matcher_test.h:74]: Same typo here and other copy-pasted constants.

File: test/extensions/matching/input_matchers/cel_matcher/cel_matcher_test.h
Comment (kyessenov) [inline test/extensions/matching/input_matchers/cel_matcher/cel_matcher_test.h:181]: You can remove source_info, it's not used.

File: test/extensions/matching/input_matchers/cel_matcher/cel_matcher_test.h
Comment (kyessenov) [inline test/extensions/matching/input_matchers/cel_matcher/cel_matcher_test.h:368]: Same, source_info is not needed in a test. You can also erase "id" fields if you want.

**Review by kyessenov — CHANGES_REQUESTED** [review]
Mostly LGTM, a few minor comments.

File: source/extensions/matching/input_matchers/cel_matcher/config.h
Comment (kyessenov) [reply source/extensions/matching/input_matchers/cel_matcher/config.h:45]: Oh, we should follow up and fix any possible violations. Probably having a single global builder is best.

File: test/extensions/matching/input_matchers/cel_matcher/cel_matcher_test.h
Comment (kyessenov) [reply test/extensions/matching/input_matchers/cel_matcher/cel_matcher_test.h:10]: I think it's just ParseFromString that will take string_view shortly when we upgrade protobuf. Don't have a strong opinion here.

File: test/extensions/matching/input_matchers/cel_matcher/cel_matcher_test.h
Comment (kyessenov) [reply test/extensions/matching/input_matchers/cel_matcher/cel_matcher_test.h:181]: Alright, that's reasonable. The compiler doesn't seem to read this fields at all but may in the future.

**Review by kyessenov — APPROVED** [review]
LGTM.

**Review by RyanTheOptimist — APPROVED** [review]
/lgtm deps

**Review by yanavlasov — APPROVED** [review]

**Pull request merged**
